// @aws-sdk/s3-request-presigner/dist/cjs/presigner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3RequestPresigner = void 0;
const middleware_sdk_s3_1 = require("@aws-sdk/middleware-sdk-s3");
const constants_1 = require("./constants");
class S3RequestPresigner {
    constructor(options) {
        const resolvedOptions = {
            // Allow `signingName` because we want to support usecase of supply client's resolved config
            // directly. Where service equals signingName.
            service: options.signingName || options.service || "s3",
            uriEscapePath: options.uriEscapePath || false,
            applyChecksum: options.applyChecksum || false,
            ...options,
        };
        this.signer = new middleware_sdk_s3_1.S3SignatureV4(resolvedOptions);
    }
    async presign(requestToSign, { unsignableHeaders = new Set(), unhoistableHeaders = new Set(), ...options } = {}) {
        unsignableHeaders.add("content-type");
        // S3 requires SSE headers to be signed in headers instead of query
        // See: https://github.com/aws/aws-sdk-js-v3/issues/1576
        Object.keys(requestToSign.headers)
            .map((header) => header.toLowerCase())
            .filter((header) => header.startsWith("x-amz-server-side-encryption"))
            .forEach((header) => {
            unhoistableHeaders.add(header);
        });
        requestToSign.headers[constants_1.SHA256_HEADER] = constants_1.UNSIGNED_PAYLOAD;
        if (!requestToSign.headers["host"]) {
            requestToSign.headers.host = requestToSign.hostname;
            if (requestToSign.port) {
                requestToSign.headers.host = `${requestToSign.headers.host}:${requestToSign.port}`;
            }
        }
        return this.signer.presign(requestToSign, {
            expiresIn: 900,
            unsignableHeaders,
            unhoistableHeaders,
            ...options,
        });
    }
}
exports.S3RequestPresigner = S3RequestPresigner;
